### PR TITLE
Redo swapchain with ImageBitmap

### PR DESCRIPTION
**Pull request by kainino0x**

### CODE REVIEW

**Comment by kainino0x** [comment]
@kangz @austineng Opened this PR inside my fork, for review. PTAL

**Comment by kainino0x** [comment]
I think Corentin already gave me some feedback but I can't find it.

File: spec/index.bs [hunk 1]
```diff
@@ -741,22 +741,27 @@ Canvas Rendering and Swap Chain {#swapchain}
...
 interface GPUSwapChain {
-    GPUTexture getCurrentTexture();
+    // Returns a new GPUTexture every time.
+    // There may be only 2 GPUSwapChainTextures and 1 ImageBitmap active
+    // at once.
+    // If there is more than 1 previously-returned GPUSwapChainTexture
+    // still active (not yet transferred), throws an exception.
+    // If there is any previously-returned ImageBitmap still active
+    // (not yet detached), it is detached.
+    GPUTexture getNextTexture();
```
Comment (austinEng) [inline spec/index.bs]: nit: GPUSwapChainTexture ?

**Comment by Kangz** [comment]
TBH this is a bit scary because we don't know that the image bitmap will only be used for presenting. Also does presenting the image bitmap consume it?

**Comment by kainino0x** [comment]
Yes, it consumes it:
https://developer.mozilla.org/en-US/docs/Web/API/ImageBitmapRenderingContext/transferFromImageBitmap

We could limit the ImageBitmap to only be used for this purpose at the beginning. It would still create less new API surface, I think. That said, I think it makes sense to be able to use it for other things like as a copy source in WebGPU or WebGL or 2d canvas.

File: spec/index.bs [hunk 2]
```diff
@@ -795,6 +800,11 @@ interface GPUDevice {
     GPUCommandEncoder createCommandEncoder(GPUCommandEncoderDescriptor descriptor);
 
     GPUQueue getQueue();
+
+    // Calling configureSwapChain a second time invalidates the previous one,
+    // and all of the textures it’s produced.
+    GPUSwapChain configureSwapChain(GPUSwapChainDescriptor descriptor);
+    Promise<GPUTextureFormat> getSwapChainPreferredFormat(GPUDevice device);
```
Comment (kainino0x) [inline spec/index.bs:802]: gpudevice shouldn't be here, will fix later

File: spec/index.bs [hunk 3]
```diff
@@ -795,6 +800,11 @@ interface GPUDevice {
     GPUCommandEncoder createCommandEncoder(GPUCommandEncoderDescriptor descriptor);
 
     GPUQueue getQueue();
+
+    // Calling configureSwapChain a second time invalidates the previous one,
+    // and all of the textures it’s produced.
+    GPUSwapChain configureSwapChain(GPUSwapChainDescriptor descriptor);
```
Comment (kainino0x) [inline spec/index.bs:801]: todo: do configureSwapChain and getSwapChainPreferredFormat need to know what canvas or ImageBitmapRenderingContext they're targeting? likely yes

**Comment by fserb** [comment]
Okey. I gave this a bit of thought. I think it's clever and nice to use `ImageBitmap` as the interface between WebGPU and Canvas. I like this much more than the old context solution. Issues that come to mind:

1. `ImageBitmap` can be used as this no-copy transfer (and it is in the Chrome implementation, both for `Canvas` and `OffscreenCanvas`'s `ImageBitmapRenderingContext`). But please remember that it is also an `ImageSource`, which means it needs to be able to produce an image for `Canvas2D.DrawImage` or `WebGL.texImage2D` or even as an `Image.src`. So implementations will have to be able to keep track of the texture ID for no-copy compositing, but also be able to produce an actual image, if needed. I don't think it's a deal breaker, but do keep that in mind.

2. I don't understand the requirements for `There may be only 2 GPUSwapChainTextures and 1 ImageBitmap active at once.`. Is there a technical reason for that? Can't the WebGPU produce two different textures as outputs? It feels you are trying to address an underlying problem, but I couldn't figure out what it is. In whatever case, the language around `"1 ImageBitmap active at once"` is weird. I understand the `ImageBitmap` won't be available for the `GPUSwapChainTexture`, but the `ImageBitmap` itself will still be active as long as the object exists, no?

Apart from those two issues, as I said, I think it's a nice elegant solution as a way of detaching `WebGPU` presentation. :)

**Comment by fserb** [comment]
One important thing, we are having some trouble finding people on Firefox and Safari to +1 the change that enables `ImageBitmapRenderingContext` on `OffscreenCanvas`: https://github.com/whatwg/html/pull/4455
If someone from this group could help, it would be appreciated.

**Comment by kainino0x** [comment]
> 2\. I don't understand the requirements for `There may be only 2 GPUSwapChainTextures and 1 ImageBitmap active at once.`. Is there a technical reason for that? Can't the WebGPU produce two different textures as outputs? It feels you are trying to address an underlying problem, but I couldn't figure out what it is. In whatever case, the language around `"1 ImageBitmap active at once"` is weird. I understand the `ImageBitmap` won't be available for the `GPUSwapChainTexture`, but the `ImageBitmap` itself will still be active as long as the object exists, no?

The technical reason for this is that, if the WebGPU swapchain is backed by a real system swapchain (could IIUC happen in fullscreen or with overlays/underlays or system compositor), there are strict limits on the lifecycles of the textures.

I have actually made it much more strict in the version I proposed upstream. There can now only be one swapchain image in flight instead of more. I took a look at Vulkan today and it looks like that is the limit there too. Essentially, ideally apps would only get the texture for the duration of drawing, then immediately transfer it to ImageBitmap and immediately present it.

E.g. Vulkan gives us 3 swapchain images, identified by im0, im1, im2
- We ask for a swapchain image; Vulkan tells us to use image 0 (im0).
- getNextTexture() returns a GPUTexture t0 wrapping im0.
- App draws into t0.
- transferToImageBitmap invalidates t0 and returns an ImageBitmap ib0 wrapping im0.
- transferFromImageBitmap detaches ib0 and hands it back to Vulkan.
The limitations I specified prevent us from asking for another swapchain image before Vulkan allows it.

I also changed the semantics so that instead of magically detaching the ImageBitmaps for outstanding swapchain images, instead the getNextTexture fails.

**Comment by fserb** [comment]
That makes sense.

Keep in mind that after the `transferToImageBitmap` it could be that the `ImageBitmap` needs to produce an actual image (as opposed to just present to screen).

**Comment by kainino0x** [comment]
Yup, I'm pretty sure that will be fine. All we need to do is be able to copy out of the texture. Once it's transferred to the IBRC it can be handed back to the system.

One open question is how to make it so an app can get rendering results out to an ImageBitmap without it having to go out via the IBRC.

**Pull request closed without merging**